### ISSUE TITLE
Perfomance issue with artifacts

### ISSUE THREAD

**Issue by tommasostocchialterna** [microsoft/navcontainerhelper#1243]
Hello,

We encountered a performance issue while migrating to the artifacts. In your article is shown that the container creation – once the image has already been downloaded – should took somewhere near 3 minutes.
In our tests, however, it takes 17 minutes.

This is the command we run:
```
$artifactUrl = Get-BCArtifactUrl -storageAccount "bcartifacts" -select "latest" -country "it"
New-BCContainer `
    -accept_eula `
    -containerName $containerName `
    -artifactUrl $artifactUrl `
    -Credential $credential `
    -memoryLimit 4G `
    -auth UserPassword `
    -updateHosts `
    -doNotCheckHealth `
    -accept_outdated `
    -imagename myown `
    -licenseFile "$licenseFile" `
    -additionalParameters $additionalParameters
```

**Full output of scripts**
```
NavContainerHelper is version 0.7.0.26
NavContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Datacenter - ltsc2016
Docker Client Version is 19.03.1
Docker Server Version is 19.03.1
Removing test-performance from host hosts file
Removing C:\ProgramData\NavContainerHelper\Extensions\test-performance
ArtifactUrl and ImageName specified
Fetching all docker images
Image myown:sandbox-16.4.14693.15859-it already exists
Building image myown:sandbox-16.4.14693.15859-it based on https://bcartifacts.blob.core.windows.net/sandbox/16.4.1
4693.15859/it
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
10.0.14393.3808-generic: Pulling from dynamicsnav
Generic Tag: 0.1.0.16
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3866 (ltsc2016)
Copying Platform Artifacts
Copying Database
Copying Licensefile
Copying ConfigurationPackages
Copying Extensions
Copying Applications.IT
Untagged: myown:sandbox-16.4.14693.15859-it
Deleted: sha256:cfdb1077b43da38592ed9db5e7650410a2815677cf90b5b241cf00495477c1a6
Deleted: sha256:08aa27f8300da07c42068561ea893bc94aa4c14b52f185ac7dba422b1b55d53d
Deleted: sha256:0e5151e9aa6781c9db636acb73d227f8f10568833e43623087fc6c1ab9953df3
Deleted: sha256:6f09bfb6944857f8b65f400eb4805432ce6ef230af3700b29305877043b515e9
Deleted: sha256:f7324de54471fb25ca2a15fa9e9ee3a705826f936507eea77812dbdc76f5bfec
Deleted: sha256:9bc198915300371a96d0f558f1d0c5451352406ae4852716043ee8f0c7c0b421
Deleted: sha256:1dc7712192b7f6bc3827a75e74660b4cbe62295bd0d24c20f017e8278fd5c4a7
Deleted: sha256:cc2e00d7a606d5e1bd61e28d268469a0c669b8b6dc8557f24b68fd387964ec3a
Deleted: sha256:4c4942c44c774f7ada4d5c065fe60b4691e3498bd847a92e426d00238a8ebad6
Deleted: sha256:d8154a658c2c6ad1e8b720c57960a110571d1d3fe8bab1e4230c112f2845e32f
c:\bcartifacts.cache\tmp637339540651655635
Sending build context to Docker daemon  1.755GB

Step 1/6 : FROM mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
 ---> 919a8b4f600a
Step 2/6 : ENV DatabaseServer=localhost DatabaseInstance=SQLEXPRESS DatabaseName=CRONUS IsBcSandbox=Y artifactUrl=
https://bcartifacts.blob.core.windows.net/sandbox/16.4.14693.15859/it
 ---> Running in 6d97cd519a0e
Removing intermediate container 6d97cd519a0e
 ---> 1b2530f5e1ac
Step 3/6 : COPY my /run/
 ---> f9ed7ffbab11
Step 4/6 : COPY NAVDVD /NAVDVD/
 ---> 320940bb43c9
Step 5/6 : RUN \Run\start.ps1 -installOnly
 ---> Running in a0e8b8eb1607
Using installer from C:\Run\150-new
Installing Business Central
Installing from DVD
Starting Local SQL Server
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
Starting Internet Information Server
Copying Service Tier Files
Copying PowerShell Scripts
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation
Restoring CRONUS Demo Database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Starting Business Central Service Tier
Importing CRONUS license file
Stopping Business Central Service Tier
Copying Web Client Files
Copying Client Files
Copying ModernDev Files
Copying additional files
Copying ConfigurationPackages
Copying Test Assemblies
Copying Extensions
Copying Applications
Copying Applications.IT
Installation took 227 seconds
Installation complete
Removing intermediate container a0e8b8eb1607
 ---> ca20ea98f8b7
Step 6/6 : LABEL legal="http://go.microsoft.com/fwlink/?LinkId=837447"       created="202008251008"       nav=""  
     cu=""       country="IT"       version="16.4.14693.15859"       platform="16.0.14648.15829"
 ---> Running in 171f5044c66c
Removing intermediate container 171f5044c66c
 ---> ad328a44383a
Successfully built ad328a44383a
Successfully tagged myown:sandbox-16.4.14693.15859-it
Building image took 888 seconds
Using image myown:sandbox-16.4.14693.15859-it
Disabling Health Check (always report healthy)
Creating Container test-performance
Version: 16.4.14693.15859-IT
Style: sandbox
Platform: 16.0.14648.15829
Generic Tag: 0.1.0.16
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3866 (ltsc2016)
WARNING: The container operating system matches the host operating system, but the revision is different.
If you encounter issues, you might want to specify -isolation hyperv
Using locale it-IT
Using process isolation
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\Users\bcadmin\Desktop\Alterna_D365BC_IT_ISV_PFS_PV.flf
Additional Parameters:
--volume "C:\Scambio:C:\Scambio"
Files in C:\ProgramData\NavContainerHelper\Extensions\test-performance\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container test-performance from image myown:sandbox-16.4.14693.15859-it
24b045d98451498a441ba1ada426408fbb19c13450d8ffab8f07b4b0c77b6cee
Waiting for container test-performance to be ready
Adding TEST-PERFORMANC to hosts file
Initializing...
Setting host.docker.internal to 10.0.0.4 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 10.0.0.4 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 10.0.0.1 in container hosts file (copy from host hosts file)
Starting Container
Hostname is test-performance
PublicDnsName is test-performance
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Creating Self Signed Certificate
Self Signed Certificate Thumbprint B81D58A98DF167E7F835C81104C56545616F9493
Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Enabling Financials User Experience
Using license file 'c:\run\my\license.flf'
Import License
Creating http download site
Setting SA Password and enabling SA
Creating Test as SQL User and add to sysadmin
Creating SUPER user
Container IP Address: 172.28.210.124
Container Hostname  : test-performance
Container Dns Name  : test-performance
Web Client          : http://test-performance/BC/
Dev. Server         : http://test-performance
Dev. ServerInstance : BC
Setting test-performance to 172.28.210.124 in host hosts file

Files:
http://test-performance:8080/ALLanguage.vsix

Initialization took 167 seconds
Ready for connections!
Reading CustomSettings.config from test-performance
Creating Desktop Shortcuts for test-performance
Container test-performance successfully created
```

**Comment by freddydk** [comment]
If you run that exact same script again - it should be much faster.
During the script it builds a script called myown:<tag> and next time up, it should reuse that.

**Comment by freddydk** [comment]
Could you display the output of the second run it that isn't much faster?
btw - I can see you are taking the last sandbox artifact - sandbox artifacts are published fairly often - onprem artifacts only once a month - or you can select a given version of the sandbox artifacts

**Comment by tommasostocchialterna** [comment]
Here's the output of another run as requested. As you can see, there's no improvement in performance.

```
WARNING: Container name should not exceed 15 characters
NavContainerHelper is version 0.7.0.26
NavContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Datacenter - ltsc2016
Docker Client Version is 19.03.1
Docker Server Version is 19.03.1
Removing test-performance from host hosts file
Removing C:\ProgramData\NavContainerHelper\Extensions\test-performance
ArtifactUrl and ImageName specified
Fetching all docker images
Image myown:sandbox-16.4.14693.15859-it already exists
Building image myown:sandbox-16.4.14693.15859-it based on https://bcartifacts.blob.core.windows.net/sandbox/16.4.146
93.15859/it
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
10.0.14393.3808-generic: Pulling from dynamicsnav
Digest: sha256:c49fd2d0510c4498aebfe09f0f4d466391c013aaa600a77392df32a7f1d95c08
Generic Tag: 0.1.0.16
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3866 (ltsc2016)
Copying Platform Artifacts
Copying Database
Copying Licensefile
Copying ConfigurationPackages
Copying Extensions
Copying Applications.IT
Untagged: myown:sandbox-16.4.14693.15859-it
Deleted: sha256:ad328a44383ae9260c7837d937910190e3fad414ac4e7a0fe031e5e9561e6791
Deleted: sha256:89fea7a070d675a3309a224143a41e1914fc8de41ed8ab1ffc6098733ce51bd9
Deleted: sha256:ca20ea98f8b769057c9c94638a7cff6167be6f2bea2ca30056032279333a3b95
Deleted: sha256:c2e1e771bc93dd6365b4cd9cdeeb2060a5fbdf0e4dfce60be0741aa635a57f25
Deleted: sha256:320940bb43c908e10daf357d8f1db22385ad9a7e867eb76f57084e9e352e94b3
Deleted: sha256:4c7202abccdf406208441713a3a913da4e7c44ce1f2d3b9bf1b38e3e8a94f910
Deleted: sha256:f9ed7ffbab11c766ec825355033c49f0526f1b0a01ed92a07895e5f375c42c4f
Deleted: sha256:a908eed328f3688f90c9a6854320c02614e7a37b4a664b28778b25b1f8f4743d
Deleted: sha256:1b2530f5e1acd8770cea00b0dd34cedbacabfbcd4c08a0252c075abd3b0f097c
Deleted: sha256:5d8f92abef2c2ae535664363076b89d86c6fbd64b130049509b3f27f45face44
c:\bcartifacts.cache\tmp637339572439434980
Sending build context to Docker daemon  1.755GB

Step 1/6 : FROM mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
 ---> 919a8b4f600a
Step 2/6 : ENV DatabaseServer=localhost DatabaseInstance=SQLEXPRESS DatabaseName=CRONUS IsBcSandbox=Y artifactUrl=
https://bcartifacts.blob.core.windows.net/sandbox/16.4.14693.15859/it
 ---> Running in eb83edff937d
Removing intermediate container eb83edff937d
 ---> 51f66a2d9509
Step 3/6 : COPY my /run/
 ---> 9ecc8e2e8dd2
Step 4/6 : COPY NAVDVD /NAVDVD/
 ---> aecfffb8c858
Step 5/6 : RUN \Run\start.ps1 -installOnly
 ---> Running in 426000f40a84
Using installer from C:\Run\150-new
Installing Business Central
Installing from DVD
Starting Local SQL Server
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
Starting Internet Information Server
Copying Service Tier Files
Copying PowerShell Scripts
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Determining Database Collation
Restoring CRONUS Demo Database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Starting Business Central Service Tier
Importing CRONUS license file
Stopping Business Central Service Tier
Copying Web Client Files
Copying Client Files
Copying ModernDev Files
Copying additional files
Copying ConfigurationPackages
Copying Test Assemblies
Copying Extensions
Copying Applications
Copying Applications.IT
Installation took 221 seconds
Installation complete
Removing intermediate container 426000f40a84
 ---> ea9933bbce14
Step 6/6 : LABEL legal="http://go.microsoft.com/fwlink/?LinkId=837447"       created="202008251101"       nav=""  
     cu=""       country="IT"       version="16.4.14693.15859"       platform="16.0.14648.15829"
 ---> Running in 7d2ebc32889a
Removing intermediate container 7d2ebc32889a
 ---> 40c13c2a41f5
Successfully built 40c13c2a41f5
Successfully tagged myown:sandbox-16.4.14693.15859-it
Building image took 884 seconds
Using image myown:sandbox-16.4.14693.15859-it
Disabling Health Check (always report healthy)
Creating Container test-performance
Version: 16.4.14693.15859-IT
Style: sandbox
Platform: 16.0.14648.15829
Generic Tag: 0.1.0.16
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3866 (ltsc2016)
WARNING: The container operating system matches the host operating system, but the revision is different.
If you encounter issues, you might want to specify -isolation hyperv
Using locale it-IT
Using process isolation
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\Users\bcadmin\Desktop\Alterna_D365BC_IT_ISV_PFS_PV.flf
Additional Parameters:
--volume "C:\Scambio:C:\Scambio"
Files in C:\ProgramData\NavContainerHelper\Extensions\test-performance\my:
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container test-performance from image myown:sandbox-16.4.14693.15859-it
52c9625619e05e28d4dc6b4c0bb14f0226a9d1236574b10ae797a922bdc7dc06
Waiting for container test-performance to be ready
Adding TEST-PERFORMANC to hosts file
Initializing...
Setting host.docker.internal to 10.0.0.4 in container hosts file (copy from host hosts file)
Setting gateway.docker.internal to 10.0.0.4 in container hosts file (copy from host hosts file)
Setting host.containerhelper.internal to 10.0.0.1 in container hosts file (copy from host hosts file)
Starting Container
Hostname is test-performance
PublicDnsName is test-performance
Using NavUserPassword Authentication
Starting Local SQL Server
Starting Internet Information Server
Creating Self Signed Certificate
Self Signed Certificate Thumbprint 799712764E9F4FC6FC09A943BF52A998E0C7CD9B
Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
Enabling Financials User Experience
Using license file 'c:\run\my\license.flf'
Import License
Creating http download site
Setting SA Password and enabling SA
Creating Test as SQL User and add to sysadmin
Creating SUPER user
Container IP Address: 172.28.220.190
Container Hostname  : test-performance
Container Dns Name  : test-performance
Web Client          : http://test-performance/BC/
Dev. Server         : http://test-performance
Dev. ServerInstance : BC
Setting test-performance to 172.28.220.190 in host hosts file

Files:
http://test-performance:8080/ALLanguage.vsix

Initialization took 165 seconds
Ready for connections!
Reading CustomSettings.config from test-performance
Creating Desktop Shortcuts for test-performance
Container test-performance successfully created


Days              : 0
Hours             : 0
Minutes           : 17
Seconds           : 50
Milliseconds      : 717
Ticks             : 10707177999
TotalDays         : 0,0123925671284722
TotalHours        : 0,297421611083333
TotalMinutes      : 17,845296665
TotalSeconds      : 1070,7177999
TotalMilliseconds : 1070717,7999
```

**Comment by freddydk** [comment]
What does Get-BestGenericImageName return on this machine?

**Comment by tommasostocchialterna** [comment]
Here's the result:
```
PS C:\windows\system32> Get-BestGenericImageName
mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
```

**Comment by freddydk** [comment]
That is very strange.
on your machine it should return 
```
mcr.microsoft.com/dynamicsnav:10.0.14393.3866-generic
```
try to run this:

```
    $repo = "mcr.microsoft.com/dynamicsnav"
    $tag = "*-generic"
    $imagetags = Get-BcContainerImageTags -imageName $repo
    $versions = @()
    if ($imagetags) {
        $ver = [Version]"0.0.0.0"
        $versions = $imagetags.tags | Where-Object { $_ -like $tag -and [System.Version]::TryParse($_.SubString($tag.indexOf('*'), $_.length-$tag.length+1), [ref]$ver) } | % { [System.Version]($_.SubString($tag.indexOf('*'), $_.length-$tag.length+1)) }
    }
```

and tell me what $versions is - and also $imageTags.tags
if $versions is empty, then it uses a hardcoded list, which could explain this.

**Comment by tommasostocchialterna** [comment]
Hi, both $versions and $tags are empty:
```
Versions: 

Tags: 

```

**Comment by freddydk** [comment]
So somehow, that machine is not allowed to enumerate docker images...
what happens if you open this in a browser: https://mcr.microsoft.com/v2/dynamicsnav/tags/list

**Comment by tommasostocchialterna** [comment]
Here's a partial output:
```
{
  "name": "dynamicsnav",
  "tags": [
    "10.0.13682.0",
    "10.0.13682.0-at",
    "10.0.13682.0-at-ltsc2016",
    "10.0.13682.0-at-ltsc2019",
    "10.0.13682.0-au",
    "10.0.13682.0-au-ltsc2016",
    "10.0.13682.0-au-ltsc2019",
    "10.0.13682.0-be",
    "10.0.13682.0-be-ltsc2016",
    "10.0.13682.0-be-ltsc2019",
    "10.0.13682.0-ch",
    "10.0.13682.0-ch-ltsc2016",
    "10.0.13682.0-ch-ltsc2019",
    "10.0.13682.0-cz",
    "10.0.13682.0-cz-ltsc2016",
    "10.0.13682.0-cz-ltsc2019",
    "10.0.13682.0-de",
    "10.0.13682.0-de-ltsc2016",
    "10.0.13682.0-de-ltsc2019",
    "10.0.13682.0-dk",
    "10.0.13682.0-dk-ltsc2016",
    "10.0.13682.0-dk-ltsc2019",
    "10.0.13682.0-es",
    "10.0.13682.0-es-ltsc2016",
    "10.0.13682.0-es-ltsc2019",
    "10.0.13682.0-fi",
    "10.0.13682.0-fi-ltsc2016",
    "10.0.13682.0-fi-ltsc2019",
    "10.0.13682.0-fr",
    "10.0.13682.0-fr-ltsc2016",
    "10.0.13682.0-fr-ltsc2019",
    "10.0.13682.0-gb",
    "10.0.13682.0-gb-ltsc2016",
    "10.0.13682.0-gb-ltsc2019",
    "10.0.13682.0-is",
    "10.0.13682.0-is-ltsc2016",
    "10.0.13682.0-is-ltsc2019",
    "10.0.13682.0-it",
    "10.0.13682.0-it-ltsc2016",
    "10.0.13682.0-it-ltsc2019",
    "10.0.13682.0-ltsc2016",
    "10.0.13682.0-ltsc2019",
    "10.0.13682.0-na",
    "10.0.13682.0-na-ltsc2016",
    "10.0.13682.0-na-ltsc2019"
```

If you need the full output let me know, but I think this is enough to answer your question.

**Comment by freddydk** [comment]
Not needed
Could you try to run this: 
    [Net.ServicePointManager]::SecurityProtocol = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12

in PS before running the script I suggested yesterday (to get tags)

**Comment by tommasostocchialterna** [comment]
Here's the partial output:

```
Versions: 
10.0.14393.2906 10.0.14393.2972 10.0.14393.3025 10.0.14393.3085 10.0.14393.3144 10.0.14393.3204 10.0.14393.3326 10
.0.14393.3384 10.0.14393.3443 10.0.14393.3506 10.0.14393.3630 10.0.14393.3686 10.0.14393.3750 10.0.14393.3808 10.0
.14393.3866 10.0.17134.1006 10.0.17134.1130 10.0.17134.950 10.0.17763.1040 10.0.17763.1158 10.0.17763.1217 10.0.17
763.1282 10.0.17763.1339 10.0.17763.1397 10.0.17763.437 10.0.17763.504 10.0.17763.557 10.0.17763.615 10.0.17763.67
8 10.0.17763.737 10.0.17763.864 10.0.17763.914 10.0.17763.973 10.0.18362.1016 10.0.18362.116 10.0.18362.175 10.0.1
8362.239 10.0.18362.295 10.0.18362.356 10.0.18362.476 10.0.18362.535 10.0.18362.592 10.0.18362.658 10.0.18362.720 
10.0.18362.778 10.0.18362.836 10.0.18362.900 10.0.18362.959 10.0.18363.1016 10.0.18363.476 10.0.18363.535 10.0.183
63.592 10.0.18363.658 10.0.18363.720 10.0.18363.778 10.0.18363.836 10.0.18363.900 10.0.18363.959 10.0.19041.264 10
.0.19041.329 10.0.19041.388 10.0.19041.450
Tags: 
10.0.13682.0 10.0.13682.0-at 10.0.13682.0-at-ltsc2016 10.0.13682.0-at-ltsc2019 10.0.13682.0-au 10.0.13682.0-au-lts
c2016 10.0.13682.0-au-ltsc2019 10.0.13682.0-be 10.0.13682.0-be-ltsc2016 10.0.13682.0-be-ltsc2019 10.0.13682.0-ch 1
0.0.13682.0-ch-ltsc2016 10.0.13682.0-ch-ltsc2019 10.0.13682.0-cz 10.0.13682.0-cz-ltsc2016 10.0.13682.0-cz-ltsc2019
 10.0.13682.0-de 10.0.13682.0-de-ltsc2016 10.0.13682.0-de-ltsc2019 10.0.13682.0-dk 10.0.13682.0-dk-ltsc2016 10.0.1
3682.0-dk-ltsc2019 10.0.13682.0-es 10.0.13682.0-es-ltsc2016 10.0.13682.0-es-ltsc2019 10.0.13682.0-fi 10.0.13682.0-
fi-ltsc2016 10.0.13682.0-fi-ltsc2019 10.0.13682.0-fr 10.0.13682.0-fr-ltsc2016 10.0.13682.0-fr-ltsc2019 10.0.13682.
0-gb 10.0.13682.0-gb-ltsc2016 10.0.13682.0-gb-ltsc2019 10.0.13682.0-is 10.0.13682.0-is-ltsc2016 10.0.13682.0-is-lt
sc2019 10.0.13682.0-it 10.0.13682.0-it-ltsc2016 10.0.13682.0-it-ltsc2019 10.0.13682.0-ltsc2016 10.0.13682.0-ltsc20
19 10.0.13682.0-na 10.0.13682.0-na-ltsc2016 10.0.13682.0-na-ltsc2019 10.0.13682.0-nl 10.0.13682.0-nl-ltsc2016 10.0
.13682.0-nl-ltsc2019 10.0.13682.0-no 10.0.13682.0-no-ltsc2016 10.0.13682.0-no-ltsc2019 10.0.13682.0-nz 10.0.13682.
```

**Comment by freddydk** [comment]
ok - got it.
If you run the original script twice, you should also have a better performance 2nd time.
I will check how I can ensure 1.2 in containerhelper.

**Comment by tommasostocchialterna** [comment]
I'll run the script twice and let you know the result.

**Comment by tommasostocchialterna** [comment]
Ok I confirmed this was resolved.
Here's the **first run** timing:
```
Days              : 0
Hours             : 0
Minutes           : 26
Seconds           : 35
Milliseconds      : 169
Ticks             : 15951697441
TotalDays         : 0,0184626127789352
TotalHours        : 0,443102706694444
TotalMinutes      : 26,5861624016667
TotalSeconds      : 1595,1697441
TotalMilliseconds : 1595169,7441
```
And here's the **second**:
```
Days              : 0
Hours             : 0
Minutes           : 2
Seconds           : 49
Milliseconds      : 496
Ticks             : 1694960297
TotalDays         : 0,00196175960300926
TotalHours        : 0,0470822304722222
TotalMinutes      : 2,82493382833333
TotalSeconds      : 169,4960297
TotalMilliseconds : 169496,0297
```

Thank you very much for your help.

**Comment by freddydk** [comment]
Ok, thanks.
There will be a new NavContainerHelper with this fix.
and it will be included in BcContainerHelper 1.0.4 as well - (btw. I do recommend to shift to bccontainerhelper - it might speed up things even more)
https://freddysblog.com/2020/08/11/bccontainerhelper/

**Comment by tommasostocchialterna** [comment]
Thank you for your advice. We will shift to bccontainerhelper then.
Closing the issue.
